packages/dds/tree/src/feature-libraries/indexing/anchorTreeIndex.ts

```diff
@@ -0,0 +1,285 @@
+	 */
+	public constructor(
+		private readonly forest: IForestSubscription,
+		indexer: (schemaId: TreeNodeSchemaIdentifier) => KeyFinder<TKey> | undefined,
+		private readonly getValue: (anchorNodes: TreeIndexNodes<AnchorNode>) => TValue | undefined,
+		private readonly checkTreeStatus: (node: AnchorNode) => TreeStatus | undefined,
+	) {
+		/**
+		 * Given a cursor in field mode, recursively indexes all nodes under the field.
+		 */
+		const indexField = (fieldCursor: ITreeSubscriptionCursor): void => {
+			forEachNode(fieldCursor, (nodeCursor) => {
+				const keyFinder = getOrCreate(
+					this.keyFinders,
+					fieldCursor.type,
+					// if the indexer does not return a key finder for this schema, we cache a null value to indicate the indexer
+					// does not need to be called if this schema is encountered in the future
+					(schema) => indexer(schema) ?? null,
+				);
+
+				if (keyFinder !== null) {
+					const key = keyFinder(fieldCursor);
+					const anchor = fieldCursor.buildAnchor();
+					const anchorNode = forest.anchors.locate(anchor) ?? fail("expected anchor node");
+
+					const nodes = this.nodes.get(key);
+					if (nodes !== undefined) {
+						// if the key already exists in the index, the anchor node is appended to its list of nodes
+						this.nodes.set(key, [...nodes, anchorNode]);
+					} else {
+						this.nodes.set(key, [anchorNode]);
+					}
+
+					this.anchors.set(anchorNode, anchor);
+					// when the anchor node is destroyed, delete it from the index
+					anchorNode.on("afterDestroy", () => {
+						const ns = this.nodes.get(key);
+						assert(ns !== undefined, "destroyed anchor node should be tracked by index");
+						const index = ns.indexOf(anchorNode);
+						assert(index !== -1, "destroyed anchor node should be tracked by index");
+						const newNodes = filterNodes(nodes, (n) => n !== anchorNode);
+						if (newNodes !== undefined) {
+							this.nodes.set(key, newNodes);
+						} else {
+							this.nodes.delete(key);
+						}
+						assert(
+							this.anchors.delete(anchorNode),
+							"destroyed anchor should be tracked by index",
+						);
+					});
+				}
+
+				forEachField(nodeCursor, (f) => {
+					indexField(f);
+				});
+			});
+		};
+
+		const detachedFieldKeys: FieldKey[] = [];
+		const detachedFieldsCursor = forest.getCursorAboveDetachedFields();
+		forEachField(detachedFieldsCursor, (field) => {
+			detachedFieldKeys.push(field.getFieldKey());
+		});
+
+		// index all existing trees (this includes the primary document tree and all other detached/removed trees)
+		for (const fieldKey of detachedFieldKeys) {
+			const cursor = forest.allocateCursor();
+			forest.tryMoveCursorToField({ fieldKey, parent: undefined }, cursor);
+			indexField(cursor);
+			cursor.free();
+		}
+
+		// index any new trees that are created later
+		forest.on("afterRootFieldCreated", (fieldKey) => {
```

Comment:
is this event guaranteed to fire for all content created in the tree, including modifications of value fields (e.g. user does foo.x = 5, and the x field is indexed)?

Reply:
It is not, I created a skipped test and added a todo to use a different event. Will add the proposal for it in the api doc

Reply:
wait, so are indexes not totally broken until that works? is there another forest event that is fired when that happens that we can subscribe to?

Reply:
@noencke is claiming that primitives are wrapped in nodes and so even value nodes start out detached and are moved in. this sounds vaguely familiar. have you tested this?

Reply:
ok, that's what I thought too but then I wrote that skipped test to make sure and it failed